Feature/#3 팀 썸네일 이미지 등록 api개발

sjmoon00 opened the pull request:
## 🔥 연관된 이슈

close: #3 

## 📜 작업 내용

> 팀 썸네일 이미지 등록 api 구현했습니다

## 💬 리뷰 요구사항

> 제가 테스트 할 때는 @SpringBootApplication(exclude = { SecurityAutoConfiguration.class })을 설정해 두고 테스트 했었습니다
> 더 좋은 방법이 있다면 알려주시면 감사하겠습니다!

## ✨ 기타

타이트한 일정이지만 힘냅시다!

sjmoon00 committed:
chore: file 디렉토리를 버전관리에서 제외

sjmoon00 committed:
chore: 이미지 저장용 경로 설정

sjmoon00 committed:
fix: 팀ID 타입 수정

sjmoon00 committed:
feat: 팀 썸네일 이미지 저장 구현

sjmoon00 committed:
feat: 파일 예외클래스 구현

파일용 예외클래스 구현 후 NO_IMAGE 예외 추가 후 작동 확인

sjmoon00 committed:
feat: 존재하는 팀인지 확인 후 예외발생

pykido commented on src/main/java/com/ops/ops/modules/file/api/FileController.java:
```diff
@@ -0,0 +1,29 @@
+            throw new FileException(FileExceptionType.NO_IMAGE);
+        }
+        fileService.saveThumbnail(teamId, thumbnailRequest.getImage());
+        return ResponseEntity.noContent().build();
```
현재 코드에서는 응답 코드가 204 No Content로 되어있는데, API 명세에서는 201 Created더라구요! 확인해주시면 감사할 거 같습니다-!

sjmoon00 replied:
검토 감사합니다
수정하겠습니다!

pykido commented on src/main/java/com/ops/ops/modules/file/api/FileController.java:
```diff
@@ -0,0 +1,29 @@
+    private final FileService fileService;
+
+    @PostMapping("/teams/{teamId}/image/thumbnail")
+    public ResponseEntity saveThumbnailImage(@PathVariable("teamId") Long teamId, ThumbnailRequest thumbnailRequest)
```
혹시 이 부분 @RequestPart("image") MultipartFile image으로 바로 받지 않는 이유가 있을까요? 또한 명세에는 imags라는다중 업로드가 가능하다고 되어있던데 지금은 단일 이미지만 받고 있는 거 같아 확인 부탁드립니다!

sjmoon00 replied:
원인은 모르겠지만 명세가 최종검토 이전으로 되어 있었습니다
명세에서 썸네일 이미지는 하나만 업로드 가능합니다

pykido reviewed (commented):
리뷰 한번 체크 부탁드립니다! 썸네일 이미지 등록이니 대표 이미지 1장 등록이 맞는 거 같은데 명세에는 다중 이미지 등록이 가능하더라구요?!

pykido reviewed (approved):
![funny gif](https://media0.giphy.com/media/v1.Y2lkPTc5MGI3NjExaWhtOHkxdnBsZzFpMnBibTR5OHg4YmE0djg4dmh0aDh2ZjRhYmxnZCZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/1dagNhv8Oqu6l8U3ZK/giphy.gif)
수고하셨습니다~!!

JJimini commented on src/main/java/com/ops/ops/modules/file/api/FileController.java:
```diff
@@ -0,0 +1,29 @@
+
+    private final FileService fileService;
+
+    @PostMapping("/teams/{teamId}/image/thumbnail")
+    public ResponseEntity saveThumbnailImage(@PathVariable("teamId") Long teamId, ThumbnailRequest thumbnailRequest)
```
PathVariable에 따로 teamId를 명시하지 않아도 자동으로 인식됩니다! 그리고 `ResponseEntity<Void>`로 작성해주시면 warning 안뜰겁니당

sjmoon00 replied:
넵 수정하겠습니다!

JJimini commented on src/main/java/com/ops/ops/modules/file/api/FileController.java:
```diff
@@ -0,0 +1,29 @@
+    @PostMapping("/teams/{teamId}/image/thumbnail")
+    public ResponseEntity saveThumbnailImage(@PathVariable("teamId") Long teamId, ThumbnailRequest thumbnailRequest)
+            throws IOException {
+        if (thumbnailRequest.getImage() == null || thumbnailRequest.getImage().isEmpty()) {
+            throw new FileException(FileExceptionType.NO_IMAGE);
+        }
```
예외를 던지는 로직인 해당 코드는 service에서 진행하면 좋을 것 같아요

sjmoon00 replied:
간단한 검증은 서비스까지 내릴 필요가 없을 것 같아서 컨트롤러에서 진행했었습니다
서비스에서 검증하도록 헐게요

JJimini commented on src/main/java/com/ops/ops/modules/file/application/FileService.java:
```diff
@@ -0,0 +1,59 @@
+
+    public void saveThumbnail(Long teamId, MultipartFile file) throws IOException {
+
+        String originalFilename = file.getOriginalFilename() != null ? file.getOriginalFilename() : "defaultName";
```
file의 filename은 null이 되지 않기 때문에 사용하지 않는 조건문입니다

sjmoon00 replied:
메서드를 보면 @Nullable 이기도 하고 워닝도 있어서 추가했었습니다!

JJimini replied:
아하 그렇던가요..! 해당 코드(`file.getOriginalFilename() != null `)가 always true라고 warning이 떠서 남겨두었던 comment입니다!

JJimini commented on src/main/java/com/ops/ops/modules/file/application/FileService.java:
```diff
@@ -0,0 +1,59 @@
+        String fullPath = getFullPath(saveThumbnailName);
+
+        verifyTeamExists(teamId);
+
+        file.transferTo(new java.io.File(fullPath));
```
이 부분은 file 코드 완성 시 수정해주세요!

sjmoon00 replied:
File 엔티티와 자바의 File 클래스 이름이 겹쳐서 그런 것 같습니다
엔티티 이름을 바꿔야 할까요?

JJimini commented on src/main/java/com/ops/ops/modules/file/application/dto/ThumbnailRequest.java:
```diff
@@ -0,0 +1,13 @@
+import org.springframework.web.multipart.MultipartFile;
+
+@Getter
+@Setter
```
객체의 무결성, 상태 변화 책임 명확화 등의 이유로 모든 코드에 setter는 열지 않는게 일반적입니다. 그리고 request와 response는 record로 작성하시면 getter도 열지 않아도 됩니다.

sjmoon00 replied:
넵 수정하겠습니다

JJimini commented on src/main/java/com/ops/ops/modules/file/exception/FileExceptionType.java:
```diff
@@ -0,0 +1,27 @@
+import org.springframework.http.HttpStatus;
+
+public enum FileExceptionType implements BaseExceptionType {
+    NO_IMAGE(HttpStatus.BAD_REQUEST, "No Image"),
+    TEAM_NOT_FOUND(HttpStatus.NOT_FOUND, "Team Not Found"),
```
메시지는 프론트 분들에게 전달될 내용이라서 한글로 적어주시면 더 읽기 좋을 것 같아요.
그리고 영어는 동시부터 적어주시는게 일반적인 관례이고, 팀을 찾을 수 없는 것은 TeamExceptionType에 들어가야합니다.

`ex. NOT_FOUND_TEAM(HttpStatus.NOT_FOUND, "팀을 찾을 수 없습니다.")`

sjmoon00 replied:
넵 한글로 다시 작성하겠습니다!
TeamExceptionType에 다시 작성하겠습니다

JJimini reviewed (changes requested):
수고하셨습니다~! 다만 해당 내용은 "팀"에 관한 내용이므로 teamController, teamService 등 team에 작성하시면 됩니다. file 로직이 완성이 안되어서 임시로 해야하는 곳이 많네요ㅠㅠ 최대한 빨리 개발해두겠습니다..! 

+) 코드 위치변경이 꼭 필요할 것 같아서 reject 걸어두겠습니다..!
+) @SpringBootApplication(exclude = { SecurityAutoConfiguration.class }) 이 설정은 일단은 그대로 두고 테스트하시고 제가 로그인 회원가입 개발하면서 지울게욥

sjmoon00 committed:
fix: 팀 썸네일 등록 API 를 팀 도메인으로 이동

sjmoon00 committed:
fix: 저장 로직 수정

FileCommandService 로 수정.
이미지 존재여부 검증 서비스로 이동.
리퀘스트 객체 레코드로 변경.

sjmoon00 committed:
fix: 저장 로직 수정

FileCommandService 로 수정.
이미지 존재여부 검증 서비스로 이동.
리퀘스트 객체 레코드로 변경.
팀 예외를 팀 도메인으로 이동.

sjmoon00 committed:
chore: 스프링 시큐리티 실행 제외

sjmoon00 committed:
Merge branch 'develop' into feature/#3-팀_썸네일_이미지_등록_API개발

JJimini commented on src/main/java/com/ops/ops/modules/team/api/TeamController.java:
```diff
@@ -1,9 +1,28 @@
+    private final FileCommandService fileCommandService;
+
+    @PostMapping("/teams/{teamId}/image/thumbnail")
+    public ResponseEntity<Void> saveThumbnailImage(@PathVariable("teamId") Long teamId, ThumbnailRequest thumbnailRequest)
+            throws IOException {
```
이 부분도 수정하면 좋을 것 같아요!

JJimini reviewed (approved):
수고하셨어용! controller와 service를 모두 team으로 옮겨달라는 말이었답니다😃 
썸네일, 프리뷰처럼 팀의 일부분으로 동작한다면, TeamController & Service가 책임지는 것이 명확하니까요!

+) 추가로 사용되지 않는 Exception, import(<- Setting option에서 설정할 수 있습니다!)는 정리해주세요!
+) 제가 저녁에 pr 확인을 다시 못할 수 있어서 미리 approve 해둘게요! 꼭 수정 후 머지해주세요

sjmoon00 committed:
chore: PathVariable 수정

sjmoon00 committed:
refactor: 파일 도메인에서 팀 도메인으로 이동

sjmoon00 committed:
Merge remote-tracking branch 'origin/develop' into feature/#3-팀_썸네일_이미지_등록_API개발
